sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -946,6 +947,45 @@ object TransposeWindow extends Rule[LogicalPlan] {
   }
 }
 
+/**
+ * Replaces duplicate window expressions with an alias in a Project above the Window node.
+ */
+object RemoveDuplicateWindowExprs extends Rule[LogicalPlan] {
```

Comment:
Shall we do the same thing for project and aggregate? Another big idea is to re-implement subexpression elimination in the plan level.